oximeter/db/src/oxql/ast/grammar.rs

```diff
@@ -279,7 +279,26 @@ peg::parser! {
         pub rule string_literal() -> Literal
             = s:string_literal_impl() { Literal::String(s) }
 
-        pub(super) rule integer_literal_impl() -> i128
+        pub(super) rule hex_integer_literal_impl() -> i128
+            = n:$("-"? "0x" ['0'..='9' | 'a'..='f' | 'A'..='F']+ !['.'])
+        {?
+            let Some((maybe_sign, digits)) = n.split_once("0x") else {
+                return Err("hex literals should start with '0x'");
+            };
+            let sign = if maybe_sign == "-" { -1 } else { 1 };
+            let Ok(x) = i128::from_str_radix(digits, 16) else {
+                return Err("invalid hex literal");
+            };
+            if x < i128::from(i64::MIN) {
+                Err("negative overflow")
```

Comment:
i don't think this will actually reject `-0x8000000000000000` (or lower) because `x` is the without-sign hex literal here.

probably better to write the `integer_literal_impl` rule as something like
```
maybe_sign:"-"? n:(hex_integer_literal_impl() / dec_integer_literal_impl()) {
    let sign = if maybe_sign == "-" { -1 } else { 1 };

    if n > i128::from(i64::MIN).abs() {
        Err("negative overflow")
    } else if n > i128::from(u64::MAX) {
        Err("positive overflow")
    } else {
        Ok(sign * n)
    }
}
```
?

(i first noticed this because having bounds `[i64::MIN, u64::MAX]` seemed odd, but i realize now that it's to be able to parse a literal that can be compared either in signed or unsigned contexts. if someone does `x < u64::MAX` .. well, that's not the parser's problem)

((also might be good to swap `sign * n` for something like `sign.checked_mul(n)` in case something could produce `i128::MIN` before re-signing the number. the debug-build-panic-on-overflow isn't reachable as-is but it's very load-bearing on `i128::from_str_radix` returning at most `i128::MAX`, which won't overflow with `-1`))

Reply:
Oh, nice catch thanks! I agree that this isn't doing the right thing because we're missing the sign there. I like the suggestion you made about the rule organization, I'll give it a whirl. And +1 to `checked_mul()`, always safer.

Reply:
Done in 6dead8f03